### PR TITLE
HSEARCH-3774 Improve the timeout implementation for Elasticsearch

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3774

~~Based on https://github.com/hibernate/hibernate-search/pull/2333 that must be merged first.~~ done!

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ElasticsearchClientImpl.java [hunk 1]
```diff
@@ -131,10 +131,10 @@ public void onFailure(Exception exception) {
...
 				);
 
-		long currentTimeoutValue = ( elasticsearchRequest.timeoutValue() == null ) ?
-				globalTimeoutValue : elasticsearchRequest.timeoutValue();
-		TimeUnit currentTimeoutUnit = ( elasticsearchRequest.timeoutUnit() == null ) ?
-				globalTimeoutUnit : elasticsearchRequest.timeoutUnit();
+		long currentTimeoutValue = ( elasticsearchRequest.timeoutManager() == null ) ?
+				globalTimeoutValue : elasticsearchRequest.timeoutManager().remainingTimeToHardTimeout();
+		TimeUnit currentTimeoutUnit = ( elasticsearchRequest.timeoutManager() == null ) ?
+				globalTimeoutUnit : TimeUnit.MICROSECONDS;
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ElasticsearchClientImpl.java:137]: I believe you meant `MILLISECONDS`?

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ElasticsearchClientImpl.java:137]: you're right

**Comment by fax4ever** [comment]
Is there anything else we can do to improve this contribution? Thanks

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CountWork.java [hunk 2]
```diff
@@ -87,7 +87,7 @@ protected ElasticsearchRequest buildRequest() {
 
 			if ( timeoutManager.hasHardTimeout() ) {
 				// set timeoutValue and timeoutUnit only for hard timeout
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CountWork.java:89]: I think this comment is obsolete. Maybe remove it?

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CountWork.java:89]: Sure

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CountWork.java:89]: I rephrased into:
> // soft timeout has no meaning for a count work

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CountWork.java:89]: It should be more expressive... Or at least I hope so

**Comment by yrodiere** [comment]
Merged, thanks

**Pull request merged**